File: frontend/src/features/dashboard/types/usageLive.ts
export interface DeviceUsageLiveItem {
  device_id: number | null;
  vpn_device_id: string;
  client_ip: string | null;
  recorded_at: string;
  interval_sec: number;
  rx_bytes: number;
  tx_bytes: number;
  delta_rx_bytes: number;
  delta_tx_bytes: number;
  rx_mib_per_sec: number;
  tx_mib_per_sec: number;
  total_mib_per_sec: number;
}

export interface DeviceUsageLiveResponse {
  items: DeviceUsageLiveItem[];
  max_age_sec: number;
}

export interface ClientBandwidth {
  rx_mib_per_sec: number;
  tx_mib_per_sec: number;
  total_mib_per_sec: number;
  delta_rx_bytes: number;
  delta_tx_bytes: number;
  recorded_at: string;
}

File: backend/app/features/vpn/schemas/usage_live.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class DeviceUsageLiveItem(BaseModel):
    """Latest VPN tunnel throughput sample for dashboard Live Clients."""

    device_id: Optional[int] = None
    vpn_device_id: str
    client_ip: Optional[str] = None
    recorded_at: datetime
    interval_sec: float = Field(gt=0)
    rx_bytes: int = Field(ge=0)
    tx_bytes: int = Field(ge=0)
    delta_rx_bytes: int = Field(ge=0)
    delta_tx_bytes: int = Field(ge=0)
    rx_mib_per_sec: float = Field(ge=0)
    tx_mib_per_sec: float = Field(ge=0)
    total_mib_per_sec: float = Field(ge=0)


class DeviceUsageLiveResponse(BaseModel):
    items: list[DeviceUsageLiveItem]
    max_age_sec: int
